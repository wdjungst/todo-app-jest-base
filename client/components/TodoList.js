import React from 'react';
import TodoForm from './TodoForm';
import Todo from './Todo';

class TodoList extends React.Component {
  constructor(props) {
    super(props);
    this.state = { todos: [] }
  }

  componentDidMount() {
    $.ajax({
      url: '/api/todos',
      type: 'GET',
      dataType: 'JSON'
    }).done( todos => {
      this.setState({ todos });
    });
  }

  onAdd = (text) => {
    $.ajax({
      url: '/api/todos',
      type: 'POST',
      dataType: 'JSON',
      data: { text }
    }).done( todo => {
      this.setState({ todos: [todo, ...this.state.todos] });
    }).fail( err => {
      alert(err.responseJSON.errors.text.message)
    });
  }

  onDelete = (id) => {
    $.ajax({
      url: `/api/todos/${id}`,
      type: 'DELETE',
      dataType: 'JSON'
    }).done( () => {
      this.setState({ todos: this.state.todos.filter( t => t._id !== id ) });
    });
  }

  onUpdate = (id, completed) => {
    $.ajax({
      url: `/api/todos/${id}`,
      type: 'PUT',
      dataType: 'JSON',
      data: { completed }
    }).done( todo => {
      this.setState({ 
        todos: this.state.todos.map( t => {
          if (t._id !== id)
            return t
          return todo
        })
      });
    });
  }

  render() {
    let todos = this.state.todos.map( todo => {
      return (<Todo key={todo._id} onUpdate={this.onUpdate} onDelete={this.onDelete} {...todo} />)
    });

    return (
      <div>
        <TodoForm onAdd={this.onAdd} />
        { todos.length ? 
          <ul className="collection">
            {todos}
          </ul> :
          <p className="center">Add Some items</p>
        }
      </div>
    )
  }
}

export default TodoList;
