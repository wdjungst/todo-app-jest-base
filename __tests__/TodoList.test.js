import React from 'react';
import TodoList from '../client/components/TodoList';
import renderer from 'react-test-renderer';
import { spy, fakeServer } from 'sinon';
import { mount, shallow } from 'enzyme';

describe('<TodoList />', () => {
  test('Component is rendered', () => {
    const component = renderer.create(
      <TodoList />
    );
  
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls componentDidMount() lifecycle method', () => {
    const componentDidMountSpy = spy(TodoList.prototype, 'componentDidMount');
    const component = mount(<TodoList />);
    expect(TodoList.prototype.componentDidMount.calledOnce);
    componentDidMountSpy.restore();
   });

  it('updates state on add', () => {
    const component = mount(<TodoList />);
    const server = fakeServer.create();
    server.respondWith('POST', '/api/todos', [
      200,
      {
        'Content-Type': 'application/json',
        'Content-Length': 2
      },
      '{ "text": "Buy Milk", "completed": false, "_id": 1 }'
    ]);
    component.instance().onAdd('foo') 
    server.respond();
    component.update();
    expect(component.state('todos').length).toBe(1);
    server.restore();
  });

  describe('update and delete', () => {
    let todos;
    let component;
    let server;

    beforeEach(() => {
      todos = [{ _id: 1, text: 'buy milk', completed: false }]
      component = mount(<TodoList />);
      server = fakeServer.create();
    });

    afterEach(() => {
      server.restore();
    });

    it('onUpdate updates state', () => {
      server.respondWith('PUT', '/api/todos/1', [
        200,
        {
          'Content-Type': 'application/json',
          'Content-Length': 2
        },
        '{ "text": "Buy Milk", "completed": true, "_id": 1 }'
      ]);
      component.setState({ todos });
      expect(component.state('todos')[0].completed).toBe(false);
      component.instance().onUpdate(1, true) 
      server.respond();
      component.update();
      expect(component.state('todos')[0].completed).toBe(true);
    });

    it('onDelete updates state', () => {
      server.respondWith('DELETE', '/api/todos/1', [
        200,
        {
          'Content-Type': 'application/json',
          'Content-Length': 2
        },
        '{ "text": "Buy Milk", "completed": true, "_id": 1 }'
      ]);
      component.setState({ todos });
      expect(component.state('todos').length).toBe(1)
      component.instance().onDelete(1) 
      server.respond();
      component.update();
      expect(component.state('todos').length).toBe(0)
    });
  });
});
