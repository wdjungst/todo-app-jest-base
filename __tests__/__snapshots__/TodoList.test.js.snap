exports[`<TodoForm /> Component is rendered 1`] = `
<form
  className="center"
  onSubmit={[Function]}>
  <input />
</form>
`;

exports[`<TodoList /> Component is rendered 1`] = `
<div>
  <form
    className="center"
    onSubmit={[Function]}>
    <input />
  </form>
  <p
    className="center">
    Add Some items
  </p>
</div>
`;
